src/devices/src/virtio/block/device.rs

```diff
@@ -332,886 +332,641 @@ impl VirtioDevice for Block {
+        assert_eq!(vq.used.ring[0].get().len, 0);
+    }
+
+    #[test]
+    fn test_request_execute_failures() {
+        let mut block = default_block();
+        let mem = block.mem.clone();
+        let vq = VirtQueue::new(GuestAddress(0), &mem, 16);
+        block.set_queue(0, vq.create_queue());
+        block.set_device_activated(true);
+        initialize_virtqueue(&vq);
+
+        let status_addr = GuestAddress(vq.dtable[2].addr.get());
+
+        {
+            // First descriptor no longer writable.
+            vq.dtable[0].flags.set(VIRTQ_DESC_F_NEXT);
+            vq.dtable[1].flags.set(VIRTQ_DESC_F_NEXT);
+
+            // Generate a seek execute error caused by a very large sector number.
+            mem.write_obj_at_addr::<u32>(VIRTIO_BLK_T_OUT, GuestAddress(0x1000))
+                .unwrap();
+            mem.write_obj_at_addr::<u64>(0x000f_ffff_ffff, GuestAddress(0x1000 + 8))
+                .unwrap();
+
+            invoke_handler_for_queue_event(&mut block);
+
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+            assert_eq!(vq.used.ring[0].get().len, 1);
+            assert_eq!(
+                mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_IOERR
+            );
+        }
+
+        {
+            // Reset the queue to reuse descriptors and memory.
+            vq.used.idx.set(0);
+            block.set_queue(0, vq.create_queue());
+
+            vq.dtable[1]
+                .flags
+                .set(VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE);
+            // Set sector to a valid number large enough that the full 0x1000 read will fail.
+            mem.write_obj_at_addr::<u32>(VIRTIO_BLK_T_IN, GuestAddress(0x1000))
+                .unwrap();
+            mem.write_obj_at_addr::<u64>(10, GuestAddress(0x1000 + 8))
+                .unwrap();
+
+            invoke_handler_for_queue_event(&mut block);
+
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+            assert_eq!(vq.used.ring[0].get().len, 1);
+            assert_eq!(
+                mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_IOERR
+            );
+        }
+    }
+
+    #[test]
+    fn test_unsupported_request_type() {
+        let mut block = default_block();
+        let mem = block.mem.clone();
+        let vq = VirtQueue::new(GuestAddress(0), &mem, 16);
+        block.set_queue(0, vq.create_queue());
+        block.set_device_activated(true);
+        initialize_virtqueue(&vq);
+
+        let status_addr = GuestAddress(vq.dtable[2].addr.get());
+
+        // Set sector to 0.
+        mem.write_obj_at_addr::<u64>(0, GuestAddress(0x1000 + 8))
+            .unwrap();
+        // Currently only 0, 1, 4, 8 are supported.
+        // Generate an unsupported request.
+        mem.write_obj_at_addr::<u32>(16, GuestAddress(0x1000))
+            .unwrap();
```

Comment:
This can be re-written with `RequestHeader` by adding a constructor to the struct that receives request_type and sector which are the only used fields.

We can then write RequestHeader to memory directly without needing to use offsets.

Reply:
Done.